### PR TITLE
Fix: Chain switching (from halted chain)

### LINKED ISSUE #757: `no engine instance available` error
Occurred during testing with high CPU load on the machine.

```
2024-02-07T17:02:51Z    DEBUG   Protocol.Commitments.Slot233.33         created (id=CommitmentID(0x944991abb9149af2b620d0e6c79970b3055653c6fdd2b157bfa0943ca0ab3008e9000000:233))
2024-02-07T17:02:51Z    DEBUG                                           SlotCommitmentReceived: CommitmentID(0x944991abb9149af2b620d0e6c79970b3055653c6fdd2b157bfa0943ca0ab3008e9000000:233) 
2024-02-07T17:02:51Z    DEBUG                                           SlotCommitmentRequestReceived: CommitmentID(0x944991abb9149af2b620d0e6c79970b3055653c6fdd2b157bfa0943ca0ab3008e9000000:233) 
2024-02-07T17:02:51Z    DEBUG                                           SlotCommitmentReceived: CommitmentID(0x944991abb9149af2b620d0e6c79970b3055653c6fdd2b157bfa0943ca0ab3008e9000000:233) 
[signal SIGSEGV: segmentation violation code=0x1 addr=0x110 pc=0x11bd350]

panic({0x1f3c640?, 0x4299360?})
2024-02-07T17:02:51Z    DEBUG                                           SlotCommitmentReceived: CommitmentID(0x944991abb9149af2b620d0e6c79970b3055653c6fdd2b157bfa0943ca0ab3008e9000000:233) 
2024-02-07T17:02:51Z    ERROR   Protocol                                network error (peer=12D3KooWBVt5Arqx8qhCfCfK7hv7AjfYvYTYzbT33fb4udd3JGsv error=failed to deserialize slot commitment: failed to get API for version 3: no engine instance available)
        /usr/local/go/src/runtime/panic.go:261
        /usr/local/go/src/runtime/signal_unix.go:861 +0x378 fp=0xc015445b48 sp=0xc015445ae8 pc=0x4d0db8
        /scratch/components/inx/server_utxo.go:66 +0x1b0 fp=0xc015445c70 sp=0xc015445b48 pc=0x11bd350
        /scratch/components/inx/server_utxo.go:122 +0x17 fp=0xc015445ca0 sp=0xc015445c70 pc=0x11bd697
        /scratch/components/inx/server_utxo.go:220 +0x32c fp=0xc015445dc8 sp=0xc015445ca0 pc=0x11beeac
        /scratch/components/inx/server_utxo.go:325 +0x26 fp=0xc015445e20 sp=0xc015445dc8 pc=0x11be646
        /scratch/components/inx/rangesend.go:111 +0x14c fp=0xc015445e70 sp=0xc015445e20 pc=0x11c0dac
        /scratch/components/inx/server_utxo.go:340 +0x68 fp=0xc015445ee8 sp=0xc015445e70 pc=0x11be548
        /go/pkg/mod/github.com/iotaledger/hive.go/runtime@v0.0.0-20240126143305-9caf79103e85/event/events.go:86 +0x1f fp=0xc015445f00 sp=0xc015445ee8 pc=0x104fcff
        /go/pkg/mod/github.com/iotaledger/hive.go/runtime@v0.0.0-20240126143305-9caf79103e85/workerpool/task.go:48 +0x72 fp=0xc015445f20 sp=0xc015445f00 pc=0x98c832
        /go/pkg/mod/github.com/iotaledger/hive.go/runtime@v0.0.0-20240126143305-9caf79103e85/workerpool/workerpool.go:226 +0x25 fp=0xc015445f90 sp=0xc015445f20 pc=0x98da05
        /go/pkg/mod/github.com/iotaledger/hive.go/runtime@v0.0.0-20240126143305-9caf79103e85/workerpool/workerpool.go:206 +0x50 fp=0xc015445fc8 sp=0xc015445f90 pc=0x98d930
        /go/pkg/mod/github.com/iotaledger/hive.go/runtime@v0.0.0-20240126143305-9caf79103e85/workerpool/workerpool.go:198 +0x25 fp=0xc015445fe0 sp=0xc015445fc8 pc=0x98d8a5
        /usr/local/go/src/runtime/asm_amd64.s:1650 +0x1 fp=0xc015445fe8 sp=0xc015445fe0 pc=0x4efd81
        /go/pkg/mod/github.com/iotaledger/hive.go/runtime@v0.0.0-20240126143305-9caf79103e85/workerpool/workerpool.go:198 +0x25

runtime.gopark(0xba5f8abffec9a?, 0x7f7d82?, 0x40?, 0x31?, 0x8?)
runtime.goparkunlock(...)
runtime.semacquire1(0xc000688d68, 0xe0?, 0x1, 0x0, 0x0?)
sync.runtime_Semacquire(0x1?)
sync.(*WaitGroup).Wait(0xc0000b8200?)
github.com/iotaledger/hive.go/app/daemon.(*OrderedDaemon).Run(0xc000434190?)
github.com/iotaledger/hive.go/app.(*App).Run(0xc0002b68f0)
main.main()
runtime.main()
runtime.goexit()
```

### LINKED ISSUE #785: SIGSEGV Crash
When we restarted some nodes it happened that other nodes crashed with SIGSEGV.

```
2024-02-26T18:48:35Z	ERROR  	Protocol                           	network error (peer=12D3KooWLQuTQLHS3qNf4obHSj1g8bCfiToW4HgpxbtjbNsBqTZz error=failed to deserialize block: failed to retrieve API for version 3: no engine instance available)

...repeats a couple of dozen of times ...

2024-02-26T18:48:35Z	ERROR  	Protocol                           	network error (peer=12D3KooWLQuTQLHS3qNf4obHSj1g8bCfiToW4HgpxbtjbNsBqTZz error=failed to deserialize block: failed to retrieve API for version 3: no engine instance available)
2024-02-26T18:48:35Z	ERROR  	Protocol                           	network error (peer=12D3KooWLQuTQLHS3qNf4obHSj1g8bCfiToW4HgpxbtjbNsBqTZz error=failed to deserialize block: failed to retrieve API for version 3: no engine instance available)
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x160 pc=0xdba5c2]
```
Here is a bigger chunk of the log:

[log.txt.zip](https://github.com/iotaledger/iota-core/files/14445131/log.txt.zip)

### PR DESCRIPTION
**Pull request by jonastheis**
Description
---------------

This PR fixes a bug in the chain switching logic, that caused nodes to prematurely switch to a candidate if the commitments of it's own main chain halted.

The candidate was even chosen before it was properly initialized causing a runtime crash due to a nil-pointer exception when accessing the engine.

Reason
----------

When tracking the heaviest candidates we were adding commitments to all sorted sets (`HeaviestClaimedCandidate`, `HeaviestAttestedCandidate`, `HeaviestVerifiedCandidate`) whenever we received a new latest commitment for any chain (at that time just claimed weight). When a node stops producing commitments, this lead to the fact that eventually all commitments in the `HeaviestAttestedCandidate` and `HeaviestVerifiedCandidate` had the same weight, namely `0`, and a node switched to a chain without an engine.

Fixes #757, #785

### CODE REVIEW

File: pkg/protocol/chain.go [hunk 1]
```diff
@@ -161,6 +161,20 @@ func (c *Chain) Commitment(slot iotago.SlotIndex) (commitment *Commitment, exist
 	return nil, false
 }
 
+// CumulativeVerifiedWeightAt returns the cumulative verified weight at the given slot for this chain.
+func (c *Chain) CumulativeVerifiedWeightAt(slot iotago.SlotIndex) uint64 {
+	lastCommitment, exists := c.Commitment(slot)
```
Comment (piotrm50) [inline pkg/protocol/chain.go]: Rename `lastCommitment` to `commitmentAtSlot` or similar

File: pkg/protocol/network.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+			return
+		}
+
+		callback(block, src)
+	})
+}
+
+// maxTimeDrift defines the maximum time drift that is allowed for incoming blocks.
+const maxTimeDrift = 5 * time.Second
```
Comment (piotrm50) [inline pkg/protocol/network.go]: This should use the same value as PreSolidFilters for clock drift check.

File: pkg/tests/protocol_engine_switching_test.go [hunk 3]
```diff
@@ -31,6 +33,256 @@ import (
...
+
+		require.NoError(t, node0.Protocol.Engines.Main.Get().Notarization.ForceCommitUntil(20))
+		ts.IssueValidationBlockWithOptions("revive", node0,
+			mock.WithValidationBlockHeaderOptions(
+				mock.WithSlotCommitment(ts.CommitmentOfMainEngine(node0, 20).Commitment()),
+				mock.WithIssuingTime(ts.API.TimeProvider().SlotStartTime(22)),
+			),
+		)
+
+		// TODO: assert attested and verified chains
```
Comment (piotrm50) [inline pkg/tests/protocol_engine_switching_test.go]: Resolve the TODO

**Review by piotrm50 — CHANGES_REQUESTED** [review]

File: pkg/protocol/network.go (hunk 2 above)
Comment (jonastheis) [reply pkg/protocol/network.go]: I'd even reuse the same functionality if possible. Having two separate time checks with separate constants seems like a source for errors

**Review by piotrm50 — APPROVED** [review]
lgtm

**Pull request merged**